src/QsCompiler/Transformations/FunctorGeneration.cs

```diff
@@ -148,6 +152,99 @@ public override QsStatementKind OnConjugation(QsConjugation stm)
+            /// </summary>
+            public List<QsStatement> AdditionalStatements = new List<QsStatement>();
+        }
+
+        public LiftOperationCalls()
+        : base(new TransformationsState())
+        {
+            this.Statements = new LiftOperationCalls.StatementTransformation(this);
+            this.ExpressionKinds = new LiftOperationCalls.ExpressionKindTransformation(this);
+        }
+
+        public class StatementTransformation
+        : StatementTransformation<TransformationsState>
+        {
+            public StatementTransformation(SyntaxTreeTransformation<TransformationsState> parent)
+            : base(parent)
+            {
+            }
+
+            public StatementTransformation(TransformationsState parent)
+            : base(parent)
+            {
+            }
+
+            public override QsScope OnScope(QsScope scope)
+            {
+                var statements = new List<QsStatement>();
+                foreach (var statement in scope.Statements)
+                {
+                    var transformed = this.OnStatement(statement);
+                    statements.AddRange(this.SharedState.AdditionalStatements);
+                    this.SharedState.AdditionalStatements.Clear();
+                    if (!(transformed.Statement is QsStatementKind.QsExpressionStatement expr &&
+                        expr.Item.Expression == ExpressionKind.UnitValue))
+                    {
+                        // Only add statements that are not free-floating Unit, which could have
+                        // been left behind by expression transformation.
+                        statements.Add(transformed);
+                    }
+                }
+                return new QsScope(statements.ToImmutableArray(), scope.KnownSymbols);
+            }
+        }
+
+        public class ExpressionKindTransformation
+        : ExpressionKindTransformation<TransformationsState>
+        {
+            public ExpressionKindTransformation(SyntaxTreeTransformation<TransformationsState> parent)
+            : base(parent)
+            {
+            }
+
+            public ExpressionKindTransformation(TransformationsState parent)
+            : base(parent)
+            {
+            }
+
+            public override ExpressionKind OnOperationCall(TypedExpression method, TypedExpression arg)
+            {
+                // An operation in an adjoint scope must return Unit, so lift the operation to be an
+                // an additional statement and then replace it with Unit.
+                this.SharedState.AdditionalStatements.Add(
+                    new QsStatement(
+                        QsStatementKind.NewQsExpressionStatement(
+                            new TypedExpression(
+                                base.OnOperationCall(method, arg),
+                                TypeArgsResolution.Empty, // TODO(swernli): How do I determine the right argument types here?
```

Comment:
This works for the cases in the test, but doesn't seem right. How do I fill in the right arg types here? Do I need to, or is that optional? Looking at other code, it usually either fills in `Empty` like I'm doing now, or collects the types from the higher level expression, but I can't seem to get access to those data structures here.

Reply:
You may need to store the current expression in the shared state as you traverse down to this point in the syntax tree (and maybe the current statement as well), so that you can populate these fields with whatever the original values were. I would do this populating with the type arg resolutions, the return type (even though it currently always has to be unit), the inferred information, and the location. The location is important for accurately reporting diagnostics (even though these are generated statements, they represent original code). I would also copy over the statement location for the generated statements as well for the same reason, but the comment and local declarations only need to be on the original, I think.

Reply:
As a reference for something similar the concrete call graph walker has to keep track of if it is looking at expressions inside a controlled functor application: https://github.com/microsoft/qsharp-compiler/blob/79503a2a3ce31e9beeb7c65fc4d24fdc0c2cbdf4/src/QsCompiler/Transformations/CallGraph/ConcreteCallGraphWalker.cs#L276
This could be similar to how you would need to keep track of the current expression, due to expressions being possibly nested. Instead of `this.SharedState.HasControlledDependency` it would be something like `this.SharedState.CurrentExpression` and instead of being in `OnControlledApplication` it would be in an override for `OnExpression`.

Reply:
Thanks for the tip! I think that worked, and I've got a new version that keeps track of the currently processing execution.